File: recursion/fibonacci.py
/*
 * Get nth Fibonacci number.
 */
 #O(2^n) time and O(n) space
def getNthFibRec(n):
    # Write your code here.
    if n == 1:
		return 0
	elif n == 2:
		return 1
	return getNthFib(n-1)+getNthFib(n-2)

def helper(n, mem):
	if n in mem:
		return mem[n]
	mem[n] = helper(n-1,mem)+helper(n-2,mem)
	return mem[n]

#O(2^n) time and O(n) space
def getNthFibRecMem(n):
	mem = { 1:0, 2:1 }
	return helper(n,mem)

//time O(n) and space O(1)
def getNthFib(n):
	if n == 1:
		return 0
	elif n == 2:
		return 1
	first=0
	second=1
	while n>2 :
		tmp = first+second
		first=second
		second=tmp
		n -=1
	return second

File: recursion/fibonacci.go
/*
 * Get nth Fibonacci number.
 */
 package main

//time O(2^n) and space (n)
func GetNthFibRec(n int) int {
	if n == 1 {
		return 0
	} else if n == 2 {
	  return 1
	}
	return GetNthFib(n-1)+GetNthFib(n-2)
}

//time O(n) and space O(1)
func GetNthFibIter(n int) int {
	if n == 1 {
		return 0
	} else if n == 2 {
		return 1
	}
	first := 0
	second := 1
	tmp:=0
	for n>2 {
		tmp = first+second
		first = second
		second = tmp
		n -= 1
	}
	return second
}

//time O(n) and space O(n)
func helper(n int, mem map[int]int) int {
	if val, found := mem[n]; found {
		return val
	}
  mem[n]=helper(n-1, mem)+helper(n-2, mem)
	return mem[n]
}

func GetNthFibRecurMem(n int) int {
	return helper(n, map[int]int {
		1:0,
		2:1,
	})
}
